edlib/src/edlib.tpp

```diff
@@ -1417,49 +1424,43 @@ static int obtainAlignmentHirschberg(
  * @return  Alphabet as a string of unique characters, where index of each character is its value in transformed
  *          sequences.
  */
-static string transformSequences(const char* const queryOriginal, const int queryLength,
-                                 const char* const targetOriginal, const int targetLength,
-                                 unsigned char** const queryTransformed,
-                                 unsigned char** const targetTransformed) {
+template <class AlphabetType>
+static void transformSequences(const AlphabetType* const queryOriginal, const int queryLength,
+                                 const AlphabetType* const targetOriginal, const int targetLength,
+                                 unordered_map<AlphabetType, uint32_t> &alphabetIdx,
+                                 uint32_t** const queryTransformed,
+                                 uint32_t** const targetTransformed) {
     // Alphabet is constructed from letters that are present in sequences.
     // Each letter is assigned an ordinal number, starting from 0 up to alphabetLength - 1,
     // and new query and target are created in which letters are replaced with their ordinal numbers.
     // This query and target are used in all the calculations later.
-    *queryTransformed = static_cast<unsigned char *>(malloc(sizeof(unsigned char) * queryLength));
-    *targetTransformed = static_cast<unsigned char *>(malloc(sizeof(unsigned char) * targetLength));
-
-    string alphabet = "";
+    *queryTransformed = static_cast<uint32_t *>(malloc(sizeof(uint32_t) * queryLength));
+    *targetTransformed = static_cast<uint32_t *>(malloc(sizeof(uint32_t) * targetLength));
 
     // Alphabet information, it is constructed on fly while transforming sequences.
-    // letterIdx[c] is index of letter c in alphabet.
-    unsigned char letterIdx[MAX_UCHAR + 1];
-    bool inAlphabet[MAX_UCHAR + 1]; // inAlphabet[c] is true if c is in alphabet
-    for (int i = 0; i < MAX_UCHAR + 1; i++) inAlphabet[i] = false;
+    // alphabetIdx[c] is index of symbol c in alphabet.
 
+    uint32_t currentSize = 0;
     for (int i = 0; i < queryLength; i++) {
-        unsigned char c = static_cast<unsigned char>(queryOriginal[i]);
-        if (!inAlphabet[c]) {
-            inAlphabet[c] = true;
-            letterIdx[c] = static_cast<unsigned char>(alphabet.size());
-            alphabet += queryOriginal[i];
+        AlphabetType c = queryOriginal[i];
+        if (alphabetIdx.find(c) == alphabetIdx.end()) {
+            alphabetIdx[c] = currentSize;
+            currentSize++;
         }
-        (*queryTransformed)[i] = letterIdx[c];
+        (*queryTransformed)[i] = alphabetIdx[c];
```

Comment:
I am not sure if compiler will optimize code above to not actually call alphabetIdx three times with find(c) (once it is actual find(c), two times it is [c]) -> maybe we should manually ensure that?